extractors/cds/tools/index-files.js

```diff
@@ -0,0 +1,88 @@
+const { existsSync, readFileSync, statSync } = require('fs');
+const { execSync, spawnSync } = require('child_process');
+const { dirname, join, resolve } = require('path');
+
+console.log('Indexing CDS files');
+
+const responseFile = process.argv[2];
+
+const npmInstallCmdWithArgs = 'npm install --quiet --no-audit --no-fund --no-package-lock';
+
+if (!existsSync(responseFile)) {
+    console.log(`'codeql database index-files --language cds' terminated early as response file '${responseFile}' does not exist. This is because no CDS files were selected or found.`);
+    process.exit(0);
+}
+
+if (statSync(responseFile).size === 0) {
+    console.log(`'codeql database index-files --language cds' terminated early as response file '${responseFile}' is empty. This is because no CDS files were selected or found.`);
+    process.exit(0);
+}
+
+let cdsCommand = 'cds';
+try {
+    execSync('cds --version', { stdio: 'ignore' });
+} catch {
+    console.log('Pre-installing cds compiler');
+    const responseFiles = readFileSync(responseFile, 'utf-8').split('\n').filter(Boolean);
+    const packageJsonDirs = new Set();
+
+    responseFiles.forEach(file => {
+        let dir = dirname(file);
+        while (dir !== resolve(dir, '..')) {
+            if (existsSync(join(dir, 'package.json')) && readFileSync(join(dir, 'package.json'), 'utf-8').includes('@sap/cds')) {
+                packageJsonDirs.add(dir);
+                break;
```

Comment:
This is a change in behaviour from the bash script, which would also capture "grandfathered" `package.json` files - e.g. `package.json`s that occurred in a parent directory of the first `package.json` that was found. I think it's probably a reasonable change, but just wanted to note it.